compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
+        slabel_ ("Stage: "<>T.unpack stage) $ vcat $ concatMap diagSC diags
+
+    storeContents ::
+        [(FilePath, [(T.Text, [LSP.Diagnostic])])]
+        -- ^ Source File, Stage Source, Diags
+    storeContents =
+        map (\(uri, StoreItem _ si) ->
+                 (fromMaybe dontKnow $ uriToFilePath uri, getDiags si))
+            $ Map.assocs
+            $ getStore
+            $ removeEmptyStages ds
+
+    dontKnow :: IsString s => s
+    dontKnow = "<unknown>"
+
+    getDiags :: DiagnosticsBySource -> [(T.Text, [LSP.Diagnostic])]
+    getDiags = map (\(ds, diag) -> (fromMaybe dontKnow ds, toList diag)) . Map.assocs
+
+emptyDiagnostics :: Diagnostics stage
+emptyDiagnostics = Diagnostics mempty
+
+-- | Sets the diagnostics for a file and compilation step
+--   if you want to clear the diagnostics call this with an empty list
+setStageDiagnostics ::
+  Show stage =>
+  FilePath ->
+  Maybe UTCTime ->
+  -- ^ the time that the file these diagnostics originate from was last edited
+  stage ->
+  [LSP.Diagnostic] ->
+  Diagnostics stage ->
+  IO (Diagnostics stage)
+setStageDiagnostics fp timeM stage diags (Diagnostics ds) = do
+    uri <- toUri fp
+    let thisFile = Map.lookup uri ds
+        addedStage :: StoreItem
+        addedStage = StoreItem (Just posixTime) $ storeItem thisFile
+    pure $ Diagnostics $ Map.insert uri addedStage ds where
+
+    storeItem :: Maybe StoreItem -> DiagnosticsBySource
+    storeItem = \case
+        Nothing ->
+            Map.singleton k v
+
+        Just (StoreItem Nothing _) ->
+            noVersions
+
+        Just (StoreItem (Just prevTime) bySource) ->
+            case compare posixTime prevTime of
+                EQ -> insert bySource
+                GT -> insert $ Map.map (SL.map tagOutdated)  bySource
+                LT -> bySource
+
+    -- Change the color of outdated diagnostics
+    tagOutdated :: Diagnostic -> Diagnostic
+    tagOutdated d = d{_severity = Just DsInfo}
+
+    isOutdated :: Diagnostic -> Bool
+    isOutdated Diagnostic{_severity} = _severity == Just DsInfo
+
+    removeOutdated, insert :: DiagnosticsBySource -> DiagnosticsBySource
+    removeOutdated = Map.filter (not . any isOutdated)
+
+    insert bySource =
+        let removeExpired =
+                if Map.lookup k bySource == Just (SL.map tagOutdated v)
+                then removeOutdated
+                else id
+        in Map.insert k v $ removeExpired bySource
+
+    k = Just $ T.pack $ show stage
+    v = SL.toSortedList diags
+
+    posixTime :: Int
+    posixTime = maybe 0 (fromEnum . utcTimeToPOSIXSeconds) timeM
+
+noVersions :: HasCallStack => a
+noVersions =
+    error "All store items must have versions"
+
+fromUri :: LSP.Uri -> FilePath
+fromUri = fromMaybe noFilePath . uriToFilePath
+
+-- TODO (MK) We canonicalize to make sure that the two files agree on use of
+-- / and \ and other shenanigans.
+-- haskell-lsp also has serious problems with relative file paths
+toUri :: FilePath -> IO LSP.Uri
```

Comment:
This is a sufficiently bad hack, that I would prefer to inline the definition and add a comment each time we have to do this. Also as mentioned above, using different keys for diagnostics and values seems problematic. `makeAbsolute` might be more appropriate than `canonicalizePath` but doesn’t solve the problem. We should have some tests that test weird combinations of filepaths.

Reply:
Given horrible hacks, I'd rather we give it a scary name, to toUriHack, and then we can have a single place which describes all the hacks - otherwise one instance will lose the scary comment. Is canonicalize really correct? Does it work with symlinks? Or does it inadvertently chase them, thus giving wrong file errors?